File: calc_acc.py
import numpy as np
from Get_global_value import num_q
from Get_global_value import J_type
from Get_global_value import Ez
from Get_global_value import BB
from Get_global_value import cc
from cross import cross
from Get_global_value import c0


def calc_acc(A0, AA, w0, ww, vd0, wd0, q, qd, qdd):
    vd = np.zeros((num_q, 3))
    wd = np.zeros((num_q, 3))

    if num_q == 0:
        print('Single body, there is no link')
    else:
        A_I_0 = A0
        for i in range(num_q):
            if BB[i] == -1:
                A_I_i = AA[i, :, :]
                if J_type[i] == 'R':
                    wd[i, :] = wd0[0:3] + cross(ww[i, :], A_I_i @ Ez * qd[i]) + A_I_i @ Ez * qdd[i]
                    vd[i, :] = vd0[0:3] \
                               + cross(wd0[0:3], A_I_0 @ c0[i, :]) \
                               + cross(w0[0:3], cross(w0[0:3], A_I_0 @ c0[i, :])) \
                               - cross(wd[i, :], A_I_i @ cc[i, i, :]) \
                               - cross(ww[i, :], cross(ww[i, :], A_I_i @ cc[i, i, :]))

                else:
                    wd[i, :] = wd0[0:3]
                    vd[i, :] = vd0[0:3] \
                               + cross(wd0[0:3], np.dot(A_I_0, c0[i, :])) \
                               + cross(w0[i, :], cross(w0[0:3], np.dot(A_I_0, c0[i, :]))) \
                               + cross(wd[i, :], np.dot(np.dot(A_I_i, Ez), q[i])) \
                               + cross(ww[i, :], cross(ww[i, :], np.dot(np.dot(A_I_i, Ez), q[i]))) \
                               + 2 * cross(ww[i, :], np.dot(np.dot(A_I_i, Ez), qd[i])) \
                               + np.dot(np.dot(A_I_i, Ez), qdd[i]) \
                               - cross(wd[i, :], np.dot(A_I_i, cc[i, i, :])) \
                               - cross(ww[i, :], cross(ww[i, :], np.dot(A_I_i, cc[i, i, :])))

            else:
                A_I_BB = AA[BB[i], :, :]
                A_I_i = AA[i, :, :]
                if J_type[i] == 'R':
                    wd[i, :] = wd[BB[i], :] + cross(ww[i, :], np.dot(np.dot(A_I_i, Ez), qd[i])) + np.dot(np.dot(A_I_i, Ez), qdd[i])
                    vd[i, :] = vd[BB[i], :] \
                               + cross(wd[BB[i], :], np.dot(A_I_BB, cc[BB[i], i, :])) \
                               + cross(ww[BB[i], :], cross(ww[BB[i], :], np.dot(A_I_BB, cc[BB[i], i, :]))) \
                               - cross(wd[i, :], np.dot(A_I_i, cc[i, i, :])) \
                               - cross(ww[i, :], cross(ww[i, :], np.dot(A_I_i, cc[i, i, :])))
                else:
                    wd[i, :] = wd[BB[i], :]
                    vd[i, :] = vd[BB[i], :] \
                               + cross(wd[BB[i], :], np.dot(A_I_BB, cc[BB[i], i, :])) \
                               + cross(ww[BB[i], :], cross(ww[BB[i], :], np.dot(A_I_BB, cc[BB[i], i, :]))) \
                               + cross(wd[i, :], np.dot(np.dot(A_I_i, Ez), q[i])) \
                               + cross(ww[i, :], cross(ww[i, :], np.dot(np.dot(A_I_i, Ez), q[i]))) \
                               + 2 * cross(ww[i, :], np.dot(np.dot(A_I_i, Ez), qd[i])) \
                               + np.dot(np.dot(A_I_i, Ez), qdd[i]) \
                               - cross(wd[i, :], np.dot(A_I_i, cc[i, i, :])) \
                               - cross(ww[i, :], cross(ww[i, :], np.dot(A_I_i, cc[i, i, :])))

    return vd, wd

File: src/matlab/spacedyn_v2r1/calc_acc.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% The SpaceDyn, a MATLAB toolbox for Space and Mobile Robots.
% (C)1998 The Space Robotics Lab. directed by Kazuya Yoshida,
% Tohoku University, Japan.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%   version 1 // Oct 4, 1999, Last modification by K.Yoshida
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%
% CALC_ACC 	Calculate the accelation of links
%
%   CALC_ACC returns the accelation in the Inertia frame
%   for link 1 to n.
%
%
%   1997.6.13 T.Hiraoka
%   2002.2.27 H.Hamano modified to version up
%

function SV = calc_acc( LP, SV )

global Ez


% If Single body
if LP.num_q == 0
   
   SV.vd = [];
   SV.wd = [];
   
% If Multi body system
else
   
   % Calcuration of coordinate transfromation matrices
   A_I_0 = SV.A0;
   
   % Calculation of acceletion vectors vd,wd
   for i = 1 : LP.num_q
      
      % Check the link connection: Is the lower one of this link, 0 ?
      if LP.BB(i) == 0
         
         % If the i-th link connects with 0-th link
         A_I_i = SV.AA(:,i*3-2:i*3);
         
         % Rotational joint
         if LP.J_type(i) == 'R'
            SV.wd(:,i) = SV.wd0(:) ...
               + cross(SV.ww(:,i),(A_I_i*Ez*SV.qd(i))) ...
               + A_I_i*Ez*SV.qdd(i);
            SV.vd(:,i) = SV.vd0(:) ...
               + cross( SV.wd0(:),(A_I_0*LP.c0(:,i)) ) ...
               + cross( SV.w0(:),cross(SV.w0(:),(A_I_0*LP.c0(:,i))) )  ...
               - cross( SV.wd(:,i),(A_I_i*LP.cc(:,i,i)) ) ...
               - cross( SV.ww(:,i), cross(SV.ww(:,i),(A_I_i*LP.cc(:,i,i))) );
            
         % Prismatic joint
         else
            SV.wd(:,i) = SV.wd0(:);
            SV.vd(:,i) = SV.vd0(:) ...
               + cross( SV.wd0(:),(A_I_0*LP.c0(:,i)) ) ...
               + cross( SV.w0(:),cross(SV.w0(:),(A_I_0*LP.c0(:,i))) ) ...
               + cross( SV.wd(:,i),(A_I_i*Ez*SV.q(i)) ) ...
               + cross( SV.ww(:,i),cross(SV.ww(:,i),(A_I_i*Ez*SV.q(i))) ) ...
               + 2*cross(SV.ww(:,i),(A_I_i*Ez*SV.qd(i))) + (A_I_i*Ez*SV.qdd(i)) ...
               - cross( SV.wd(:,i),(A_I_i*LP.cc(:,i,i)) ) ...
               - cross( SV.ww(:,i),cross(SV.ww(:,i),(A_I_i*LP.cc(:,i,i))) ) ;
            
         end
         
         
      % Current (i-th) link doesn't have connection with the 0-th link
      else
         A_I_BB = SV.AA(:,LP.BB(i)*3-2:LP.BB(i)*3);
         A_I_i  = SV.AA(:,i*3-2:i*3);
         
         % Rotational joint
         if LP.J_type(i) == 'R'
            SV.wd(:,i) = SV.wd(:,LP.BB(i)) ...
               + cross( SV.ww(:,i),(A_I_i*Ez*SV.qd(i)) ) + (A_I_i*Ez*SV.qdd(i));
            SV.vd(:,i) = SV.vd(:,LP.BB(i)) ...
               + cross( SV.wd(:,LP.BB(i)),(A_I_BB*LP.cc(:,LP.BB(i),i)) ) ...
               + cross( SV.ww(:,LP.BB(i)),cross(SV.ww(:,LP.BB(i)),(A_I_BB*LP.cc(:,LP.BB(i),i))) ) ...
               - cross( SV.wd(:,i),(A_I_i*LP.cc(:,i,i)) ) ...
               - cross( SV.ww(:,i),cross(SV.ww(:,i),(A_I_i*LP.cc(:,i,i))) );
         
         % Prismatic joint
         else
            SV.wd(:,i) = SV.wd(:,LP.BB(i));
            SV.vd(:,i) = SV.vd(:,LP.BB(i)) ...
               + cross( SV.wd(:,LP.BB(i)),(A_I_BB*LP.cc(:,LP.BB(i),i)) ) ...
               + cross( SV.ww(:,LP.BB(i)),cross(SV.ww(:,LP.BB(i)),(A_I_BB*LP.cc(:,LP.BB(i),i))) ) ...
               + cross( SV.wd(:,i),(A_I_i*Ez*SV.q(i)) ) ...
               + cross( SV.ww(:,i),cross(SV.ww(:,i),(A_I_i*Ez*SV.q(i))) ) ...
               + 2*cross( SV.ww(:,i),(A_I_i*Ez*SV.qd(i)) ) + (A_I_i*Ez*SV.qdd(i)) ...
               - cross( SV.wd(:,i),(A_I_i*LP.cc(:,i,i)) ) ...
               - cross( SV.ww(:,i),cross(SV.ww(:,i),(A_I_i*LP.cc(:,i,i))) );
            
         end
         
      end
      
   end
   
end


%%%EOF
